File: compat/suites/python-sdk/lib/scenario/assertions.py
"""
lib/scenario/assertions.py — the closed assertion set.

compat/model/README.md § Assertions is the normative table; this module is one
function per row and nothing else. Every failure goes through
``Executor.fail``, so all six required fields are present in every message by
construction rather than by remembering.
"""

from __future__ import annotations

import re
import time
from dataclasses import dataclass
from typing import Any, Optional

from .executor import Executor, StepRef, describe_error, error_matches
from .expressions import (
    as_json_document,
    check_equals_json_operand,
    compact_json,
    is_non_empty,
    json_document_equal,
    json_equal,
    resolve_path,
)
from .failures import MISSING, ScenarioError, ScenarioFailure, clip, render, render_clipped


@dataclass
class Primary:
    """The test's own call: what a clause with no ``call`` of its own reads,
    and what fields 2 and 3 name for such a clause."""

    op: str
    params: dict
    response: Optional[dict]
    error: Optional[Exception]


def check_clause(ex: Executor, clause: dict, ref: StepRef, primary: Primary) -> None:
    """Evaluate one assertion clause. Returns on success; raises
    :class:`ScenarioFailure` with the six-field message otherwise."""
    kind = clause["kind"]
    if kind == "responseField":
        return _response_field(ex, clause, ref, primary)
    if kind == "readback":
        return _readback(ex, clause, ref)
    if kind == "listContains":
        return _list_membership(ex, clause, ref, primary, want_member=True)
    if kind == "absent":
        if "error" in clause:
            return _absent_error(ex, clause, ref)
        return _list_membership(ex, clause, ref, primary, want_member=False)
    if kind == "errorCode":
        return _error_code(ex, clause, ref, primary)
    if kind == "eventually":
        return _eventually(ex, clause, ref, primary)
    raise ScenarioFailure(
        f"{ref.group}/{ref.test}: unknown assertion kind {kind!r} "
        f"at {ref.file} {ref.step}"
    )


# ── responseField ────────────────────────────────────────────────────────────

def _response_field(ex: Executor, clause: dict, ref: StepRef, primary: Primary) -> None:
    if primary.response is None:
        # The generator refuses a test that both expects an error and reads its
        # own response, so this is a corrupt scenario file rather than a
        # failing emulator.
        raise ex.fail(
            ref=ref, op=primary.op, params=primary.params, assertion="responseField",
            expected="the test's own response, to check against",
            actual=describe_error(primary.error),
        )
    _run_checks(ex, clause["checks"], primary.response, ref, "responseField",
                primary.op, primary.params)


# ── readback ─────────────────────────────────────────────────────────────────

def _readback(ex: Executor, clause: dict, ref: StepRef) -> None:
    call = clause["call"]
    # Exports are deferred: a readback's exports are applied only when its
    # checks pass, so a failed attempt inside an `eventually` leaves the
    # context bag exactly as it found it.
    params, response = ex.perform(call, ref, "readback", apply_exports=False)
    _run_checks(ex, clause["checks"], response, ref, "readback", call["op"], params)
    ex.apply_export(call, params, response, ref, "readback")


# ── listContains / absent (list form) ────────────────────────────────────────

def _list_membership(ex: Executor, clause: dict, ref: StepRef, primary: Primary,
                     *, want_member: bool) -> None:
    kind = "listContains" if want_member else "absent"
    call = clause.get("call")
    if call is not None:
        params, response = ex.perform(call, ref, kind, apply_exports=False)
        op = call["op"]
    else:
        if primary.response is None:
            raise ex.fail(
                ref=ref, op=primary.op, params=primary.params, assertion=kind,
                expected="the test's own response, to read the list from",
                actual=describe_error(primary.error),
            )
        params, response, op = primary.params, primary.response, primary.op

    items_path = clause["itemsPath"]
    items = resolve_path(response, items_path)
    try:
        where = {path: ex.evaluate(expected)
                 for path, expected in clause["where"].items()}
    except ScenarioError as exc:
        raise ex.fail(
            ref=ref, op=op, params=params, assertion=kind, path=items_path,
            expected=f"the where clause {render(clause['where'])}, once evaluated",
            actual=str(exc),
        ) from exc
    criteria = ", ".join(f"{path}={render(value)}" for path, value in where.items())

    if items is MISSING:
        # A missing list counts as empty: several AWS services omit the member
        # rather than serialize []. Empty satisfies `absent` and fails
        # `listContains`, which needs a non-empty list containing the item.
        items = []
    elif not isinstance(items, list):
        raise ex.fail(
            ref=ref, op=op, params=params, assertion=kind, path=items_path,
            expected="a list", actual=render_clipped(items),
        )

    matched = [item for item in items if _matches(item, where)]

    if want_member and not matched:
        raise ex.fail(
            ref=ref, op=op, params=params, assertion=kind, path=items_path,
            expected=f"a non-empty list containing an item where {criteria}",
            actual=(f"{len(items)} item(s): {render_clipped(items)}" if items
                    else "an empty or absent list"),
        )
    if not want_member and matched:
        raise ex.fail(
            ref=ref, op=op, params=params, assertion=kind, path=items_path,
            expected=f"no item where {criteria}",
            actual=f"{len(matched)} matching item(s): {render_clipped(matched)}",
        )

    if call is not None:
        ex.apply_export(call, params, response, ref, kind)


def _matches(item: Any, where: dict[str, Any]) -> bool:
    """An item matches when every ``where`` entry is equal, as JSON. ``$`` is
    the item itself, which is how a list of bare strings is matched."""
    for path, expected in where.items():
        if not json_equal(resolve_path(item, path), expected):
            return False
    return True


# ── absent (error form) / errorCode ──────────────────────────────────────────

def _absent_error(ex: Executor, clause: dict, ref: StepRef) -> None:
    call = clause["call"]
    params, _response, error = ex.attempt(call, ref, "absent")
    if error is not None and error_matches(error, clause["error"]):
        return
    raise ex.fail(
        ref=ref, op=call["op"], params=params, assertion="absent",
        expected=_accepted(clause["error"]), actual=describe_error(error),
    )


def _error_code(ex: Executor, clause: dict, ref: StepRef, primary: Primary) -> None:
    if primary.error is not None and error_matches(primary.error, clause["error"]):
        return
    raise ex.fail(
        ref=ref, op=primary.op, params=primary.params, assertion="errorCode",
        expected=_accepted(clause["error"]), actual=describe_error(primary.error),
    )


def _accepted(error: dict) -> str:
    """The fifth field's "expected" half: the codes the clause accepts. Either
    of the two, matched against either the SDK's reported code or its type
    name, because the SDKs disagree about which they surface."""
    codes = sorted({error.get("shape"), error.get("code")} - {None})
    return "an error whose code or type name is one of " + ", ".join(codes)


# ── eventually ───────────────────────────────────────────────────────────────

def _eventually(ex: Executor, clause: dict, ref: StepRef, primary: Primary) -> None:
    """Retry the inner clause up to ``maxAttempts`` times, ``delayMs`` apart.

    The bounded poll loop is the sanctioned alternative to a fixed sleep: it
    only exists where the recipe declared the resource eventually consistent,
    and the budget is the recipe author's. The last attempt's failure is the
    one reported, because it is the state the service actually settled in."""
    max_attempts = clause["maxAttempts"]
    delay_s = clause.get("delayMs", 0) / 1000.0
    inner = clause["assert"]
    inner_ref = ref.child("assert")
    last: Optional[ScenarioFailure] = None
    for attempt in range(1, max_attempts + 1):
        try:
            check_clause(ex, inner, inner_ref, primary)
            return
        except ScenarioFailure as exc:
            last = exc
            if attempt < max_attempts and delay_s > 0:
                time.sleep(delay_s)
    raise ScenarioFailure(
        f"eventually gave up after {max_attempts} attempt(s) "
        f"{int(delay_s * 1000)}ms apart; last failure: {last}"
    )


# ── checks ───────────────────────────────────────────────────────────────────

def _run_checks(ex: Executor, checks: dict, response: dict, ref: StepRef,
                assertion: str, op: str, params: dict) -> None:
    for path, check in checks.items():
        value = resolve_path(response, path)
        name, argument = next(iter(check.items()))
        try:
            expected, holds, actual = _check(name, argument, value, ex)
        except ScenarioError as exc:
            # An `equals` whose expected value is a $ref nothing exported, or
            # an `equalsJSON` operand the loader should already have refused.
            raise ex.fail(
                ref=ref, op=op, params=params, assertion=assertion, path=path,
                expected=f"{name} {render(argument)}, once evaluated",
                actual=str(exc),
            ) from exc
        except re.error as exc:
            # A `matches` pattern the model states in RE2 that `re` will not
            # compile. It is a normal mismatch with the same six fields as any
            # other, never a raw exception out of the evaluator: the run must
            # report which test and which pattern, not a traceback.
            raise ex.fail(
                ref=ref, op=op, params=params, assertion=assertion, path=path,
                expected=f"pattern {argument}",
                actual=f"unsupported pattern: {exc}",
            ) from exc
        if not holds:
            raise ex.fail(
                ref=ref, op=op, params=params, assertion=assertion, path=path,
                expected=expected,
                actual=render_clipped(value) if actual is None else actual,
            )


def _check(name: str, argument: Any, value: Any,
           ex: Executor) -> tuple[str, bool, Optional[str]]:
    """One check: its "expected" description, whether it holds, and its own
    "actual" description when the plain rendering of the value is not it."""
    if name == "nonEmpty":
        return "a value that is not null, \"\", [] or {}", is_non_empty(value), None
    if name == "isList":
        # True of a list, empty or not, and true of a member the service
        # omitted rather than serializing as []. A present non-list fails.
        return ("a list, empty or absent",
                value is MISSING or isinstance(value, list), None)
    if name == "equals":
        expected = ex.evaluate(argument)
        return f"equals {render(expected)}", json_equal(value, expected), None
    if name == "equalsJSON":
        return _equals_json(argument, value)
    if name == "matches":
        # Compiled before the value is looked at, so a pattern `re` rejects is
        # reported as an unsupported pattern whatever the response held.
        pattern = re.compile(argument)
        return (f"matches /{argument}/",
                isinstance(value, str) and pattern.search(value) is not None, None)
    if name == "missing":
        return "the path not to resolve", value is MISSING, None
    raise ScenarioFailure(f"unknown check {name!r}")


def _equals_json(operand: Any, value: Any) -> tuple[str, bool, Optional[str]]:
    """``equalsJSON``: the value is a JSON document equal to the operand.

    The operand is never evaluated — it is a literal document. botocore has
    already decoded an IAM policy member into a dict, which is the document as
    it stands; a string is percent-decoded once and parsed. The actual side is
    the decoded document, so a reader compares two documents rather than a
    percent-encoded string with an object."""
    check_equals_json_operand(operand)
    expected = f"equalsJSON {compact_json(operand)}"
    if value is MISSING:
        return expected, False, render(value)
    is_document, document = as_json_document(value)
    if not is_document:
        return expected, False, f"not a JSON document: {render_clipped(value)}"
    return (expected, json_document_equal(document, operand),
            f"document {clip(compact_json(document))}")

File: compat/suites/node-js-sdk/src/lib/scenario/assertions.ts
/**
 * assertions.ts — the predicates behind the closed assertion set.
 *
 * Everything here is pure: it takes a response (or an error) that somebody
 * else fetched and says whether a clause holds, returning `null` when it does
 * and the expected/actual pair when it does not. Making the calls, retrying
 * them and applying exports is executor.ts's job — keeping the two apart is
 * what lets every clause share one failure message and lets the unit tests
 * cover the semantics without a client at all.
 *
 * The semantics are compat/model/README.md § Assertions, and the three that
 * are easy to get subtly wrong are spelled out at their implementations:
 * `isList` (absence passes), `nonEmpty` (numbers and booleans never empty)
 * and the error forms (code *or* type name, against shape *or* code).
 */

import {
  describeClipped,
  evaluateValue,
  isEmpty,
  jsonEquals,
  resolvePath,
} from "./expressions.ts";
import type { EvalContext } from "./expressions.ts";
import type { Check, Checks, ErrorSpec, Path, Where } from "./ir.ts";

/** Why a clause did not hold — fields 4 and 5 of the failure message. */
export interface Mismatch {
  /** The checks/where path, when the clause has one. */
  path?: string;
  expected: string;
  actual: string;
}

/** `actual`'s rendering — the shared clip helper, so a big page is elided. */
const render = describeClipped;

/**
 * Every check against one response, in declaration order. The first that
 * fails is the one reported.
 */
export function evaluateChecks(
  response: unknown,
  checks: Checks,
  ctx: EvalContext,
): Mismatch | null {
  for (const [path, check] of Object.entries(checks)) {
    const mismatch = evaluateCheck(response, path, check, ctx);
    if (mismatch !== null) return mismatch;
  }
  return null;
}

function evaluateCheck(
  response: unknown,
  path: Path,
  check: Check,
  ctx: EvalContext,
): Mismatch | null {
  const found = resolvePath(response, path);
  const actual = found.found ? render(found.value) : "<missing>";

  if ("missing" in check) {
    if (!found.found) return null;
    return { path, expected: "the path not to resolve", actual };
  }

  if ("isList" in check) {
    // Absence passes: a single-page List* legally returns an empty page, and
    // several AWS services (SQS's ListQueues among them) omit the member
    // entirely rather than serialize []. A present non-list still fails.
    if (!found.found) return null;
    if (Array.isArray(found.value)) return null;
    return { path, expected: "a list (or nothing)", actual };
  }

  if (!found.found) {
    return { path, expected: expectationOf(check, ctx), actual: "<missing>" };
  }

  if ("nonEmpty" in check) {
    // null, "", [] and {} are empty; numbers and booleans never are.
    if (!isEmpty(found.value)) return null;
    return { path, expected: "a non-empty value", actual };
  }

  if ("matches" in check) {
    // The pattern is compiled before the value is looked at, so what the
    // failure says does not depend on what came back: a pattern this engine
    // rejects reads "unsupported pattern" whether the path resolved to a
    // string or to a number. python and cli report it the same way, and
    // compiling second would have made one broken scenario file fail two
    // different ways across the three.
    let re: RegExp;
    try {
      re = new RegExp(check.matches);
    } catch (err) {
      // The model's patterns are RE2-compatible; a pattern legal there but not
      // in JS (a `(?P<name>…)` named group, say) must not throw out of the
      // evaluator — it has to fail the clause like any other mismatch, with
      // the same six-field message everything else gets.
      return {
        path,
        expected: `pattern ${check.matches}`,
        actual: `unsupported pattern: ${err instanceof Error ? err.message : String(err)}`,
      };
    }
    if (typeof found.value !== "string") {
      return { path, expected: `a string matching /${check.matches}/`, actual };
    }
    if (re.test(found.value)) return null;
    return { path, expected: `a string matching /${check.matches}/`, actual };
  }

  const expected = evaluateValue(check.equals, ctx);
  if (jsonEquals(found.value, expected)) return null;
  return { path, expected: render(expected), actual };
}

function expectationOf(check: Check, ctx: EvalContext): string {
  if ("nonEmpty" in check) return "a non-empty value";
  if ("matches" in check) return `a string matching /${check.matches}/`;
  // isList and missing are handled before this is reached — both hold on a
  // path that does not resolve, so neither has an expectation to render here.
  return "the path to resolve";
}

/**
 * `listContains`: the list at `itemsPath` is non-empty and contains an item
 * matching every `where` entry. A missing list fails it (it counts as empty).
 */
export function evaluateListContains(
  response: unknown,
  itemsPath: Path,
  where: Where,
  ctx: EvalContext,
): Mismatch | null {
  const items = resolveItems(response, itemsPath);
  if (items.problem !== null) return items.problem;
  if (items.list.some((item) => itemMatches(item, where, ctx))) return null;
  return {
    path: itemsPath,
    expected: `an item matching ${renderWhere(where, ctx)}`,
    actual: render(items.list),
  };
}

/**
 * `absent`, list form: no item of the list at `itemsPath` matches every
 * `where` entry. A missing list counts as empty, so it holds.
 */
export function evaluateListAbsent(
  response: unknown,
  itemsPath: Path,
  where: Where,
  ctx: EvalContext,
): Mismatch | null {
  const found = resolvePath(response, itemsPath);
  if (!found.found) return null;
  if (!Array.isArray(found.value)) {
    return { path: itemsPath, expected: "a list (or nothing)", actual: render(found.value) };
  }
  const match = found.value.find((item) => itemMatches(item, where, ctx));
  if (match === undefined) return null;
  return {
    path: itemsPath,
    expected: `no item matching ${renderWhere(where, ctx)}`,
    actual: render(match),
  };
}

function resolveItems(
  response: unknown,
  itemsPath: Path,
): { list: unknown[]; problem: Mismatch | null } {
  const found = resolvePath(response, itemsPath);
  if (!found.found) {
    return {
      list: [],
      problem: { path: itemsPath, expected: "a non-empty list", actual: "<missing>" },
    };
  }
  if (!Array.isArray(found.value)) {
    return {
      list: [],
      problem: { path: itemsPath, expected: "a list", actual: render(found.value) },
    };
  }
  return { list: found.value, problem: null };
}

/** An item matches when every `where` entry is equal, as JSON. `$` is the item. */
function itemMatches(item: unknown, where: Where, ctx: EvalContext): boolean {
  for (const [path, expectedValue] of Object.entries(where)) {
    const found = resolvePath(item, path);
    if (!found.found) return false;
    if (!jsonEquals(found.value, evaluateValue(expectedValue, ctx))) return false;
  }
  return true;
}

function renderWhere(where: Where, ctx: EvalContext): string {
  const parts = Object.entries(where).map(
    ([path, value]) => `${path}=${render(evaluateValue(value, ctx))}`,
  );
  return `{${parts.join(", ")}}`;
}

// ─── Errors ───────────────────────────────────────────────────────────────

/**
 * An error matches when its reported code *or* its type name equals the
 * clause's `shape` *or* its `code`.
 *
 * The two spellings exist because SDKs disagree about which they surface:
 * for SQS's not-found they are `QueueDoesNotExist` (the modeled shape) and
 * `AWS.SimpleQueueService.NonExistentQueue` (the awsQueryError code). The
 * JSON protocol carries whichever the service chose in `__type`, an XML one
 * carries it in the `Code` of its error node, and AWS also sends the legacy
 * one in the `x-amzn-query-error` header, so all of those surfaces are read.
 * Nothing here is conditioned on Overcast.
 */
export function errorMatches(err: unknown, spec: ErrorSpec): boolean {
  const accepted = new Set([spec.shape, spec.code]);
  return errorCodes(err).some((code) => accepted.has(code));
}

/** Every code an SDK error reports, in the spellings a clause may name. */
export function errorCodes(err: unknown): string[] {
  if (err === null || typeof err !== "object") return [];
  const e = err as Record<string, unknown>;
  const out: string[] = [];

  const add = (value: unknown): void => {
    if (typeof value !== "string" || value === "") return;
    out.push(value);
    // Smithy ids ("com.amazonaws.sqs#QueueDoesNotExist") and the header's
    // "<code>;<fault>" form both carry the bare code after a separator.
    const hash = value.lastIndexOf("#");
    if (hash >= 0) out.push(value.slice(hash + 1));
    const semi = value.indexOf(";");
    if (semi >= 0) out.push(value.slice(0, semi));
  };

  add(e["name"]);
  add(e["__type"]);
  add(e["Code"]);
  add(e["code"]);

  // The nested half of the Errors table's body-code row. An AWS Query service
  // states its code only inside the ErrorResponse envelope's error node, and
  // REST XML's bare <Error> body reads the same way wherever a parser keeps
  // the root element rather than dropping it — which is the position the
  // SDK's own loadQueryErrorCode and loadRestXmlErrorCode read. Reading only
  // the top level saw no code at all in a Query error, and every errorCode
  // clause against a Query service failed here while passing in the backends
  // that did read it (#1896).
  const nested = e["Error"];
  if (nested !== null && typeof nested === "object") {
    add((nested as Record<string, unknown>)["Code"]);
    add((nested as Record<string, unknown>)["code"]);
  }

  const response = e["$response"];
  if (response !== null && typeof response === "object") {
    const headers = (response as Record<string, unknown>)["headers"];
    if (headers !== null && typeof headers === "object") {
      add((headers as Record<string, unknown>)["x-amzn-query-error"]);
    }
  }
  return out;
}

/** "QueueDoesNotExist or AWS.SimpleQueueService.NonExistentQueue" */
export function acceptedCodes(spec: ErrorSpec): string {
  return spec.shape === spec.code ? spec.shape : `${spec.shape} or ${spec.code}`;
}
